// @salesforce/plugin-data/lib/reporters.js
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { EOL } from 'node:os';
import { Logger, Messages } from '@salesforce/core';
import { ux } from '@oclif/core';
import chalk from 'chalk';
import { get, getArray, getNumber, isString } from '@salesforce/ts-types';
import { capitalCase } from 'change-case';
import { FieldType } from './dataSoqlQueryTypes.js';
Messages.importMessagesDirectoryFromMetaUrl(import.meta.url);
const messages = Messages.loadMessages('@salesforce/plugin-data', 'soql.query');
const reporterMessages = Messages.loadMessages('@salesforce/plugin-data', 'reporter');
class Reporter {
    logger;
    constructor() {
        this.logger = Logger.childFromRoot('reporter');
    }
}
class QueryReporter extends Reporter {
    columns = [];
    data;
    constructor(data, columns) {
        super();
        this.columns = columns;
        this.data = data;
    }
}
export class HumanReporter extends QueryReporter {
    constructor(data, columns) {
        super(data, columns);
    }
    display() {
        const { attributeNames, children, aggregates } = this.parseFields();
        // in case of count() there are no records, but there is a totalSize
        const totalCount = this.data.result.records.length ? this.data.result.records.length : this.data.result.totalSize;
        this.soqlQuery(attributeNames, this.massageRows(this.data.result.records, children, aggregates), totalCount);
    }
    parseFields() {
        const fields = this.columns;
        // Field names
        const attributeNames = [];
        // For subqueries. Display the children under the parents
        const children = [];
        // For function fields, like avg(total).
        const aggregates = [];
        if (fields) {
            this.logger.info(`Found fields ${JSON.stringify(fields.map((field) => `${typeof field}.${field.name}`))}`);
            fields.forEach((field) => {
                if (field.fieldType === FieldType.subqueryField) {
                    children.push(field.name);
                    (field.fields ?? []).forEach((subfield) => attributeNames.push(`${field.name}.${subfield.name}`));
                }
                else if (field.fieldType === FieldType.functionField) {
                    if (field.alias) {
                        attributeNames.push(field.alias);
                    }
                    else {
                        attributeNames.push(field.name);
                    }
                    aggregates.push(field);
                }
                else {
                    attributeNames.push(field.name);
                }
            });
        }
        else {
            this.logger.info(`No fields found for query "${this.data.query}"`);
        }
        return { attributeNames, children, aggregates };
    }
    soqlQuery(columns, records, totalCount) {
        this.prepNullValues(records);
        ux.table(records, prepColumns(columns));
        ux.log(chalk.bold(messages.getMessage('displayQueryRecordsRetrieved', [totalCount])));
    }
    prepNullValues(records) {
        records
            .filter((record) => record)
            .forEach((record) => {
            if (record) {
                const recordAsObject = record;
                Reflect.ownKeys(recordAsObject).forEach((propertyKey) => {
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                    const value = Reflect.get(recordAsObject, propertyKey);
                    if (value === null) {
                        Reflect.set(recordAsObject, propertyKey, chalk.bold('null'));
                    }
                    else if (typeof value === 'object') {
                        this.prepNullValues([value]);
                    }
                });
            }
        });
    }
    //  public massageRows(queryResults: BasicRecord[], children: string[], aggregates: Field[]): BasicRecord[] {
    massageRows(queryResults, children, aggregates) {
        // some fields will return a JSON object that isn't accessible via the query (SELECT Metadata FROM RemoteProxy)
        // some will return a JSON that IS accessible via the query (SELECT owner.Profile.Name FROM Lead)
        // querying (SELECT Metadata.isActive FROM RemoteProxy) throws a SOQL validation error, so we have to display the entire Metadata object
        queryResults.forEach((qr) => {
            const result = qr;
            this.data.columns.forEach((col) => {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                const entry = Reflect.get(result, col.name);
                if (typeof entry === 'object' && col.fieldType === FieldType.field) {
                    Reflect.set(result, col.name, JSON.stringify(entry, null, 2));
                }
                else if (typeof entry === 'object' && col.fields?.length && entry) {
                    col.fields.forEach((field) => {
                        Reflect.set(result, `${col.name}.${field.name}`, get(result, `${col.name}.records[0].${field.name}`));
                    });
                }
            });
        });
        // There are subqueries or aggregates. Massage the data.
        if (children.length > 0 || aggregates.length > 0) {
            const qr = queryResults.reduce((newResults, result) => {
                // Aggregates are soql functions that aggregate data, like "SELECT avg(total)" and
                // are returned in the data as exprX. Aggregates can have aliases, like "avg(total) totalAverage"
                // and are returned in the data as the alias.
                if (aggregates.length > 0) {
                    for (let i = 0; i < aggregates.length; i++) {
                        const aggregate = aggregates[i];
                        if (!aggregate.alias) {
                            Reflect.set(result, aggregate.name, Reflect.get(result, `expr${i}`));
                        }
                    }
                }
                const subResults = [];
                if (children.length > 0) {
                    const childrenRows = {};
                    children.forEach((child) => {
                        const aChild = get(result, child);
                        Reflect.set(childrenRows, child, aChild);
                        Reflect.deleteProperty(result, child);
                    });
                    Reflect.ownKeys(childrenRows).forEach((child) => {
                        const childO = get(childrenRows, child);
                        if (childO) {
                            const childRecords = getArray(childO, 'records', []);
                            // eslint-disable-next-line @typescript-eslint/no-explicit-any
                            childRecords.forEach((record, index) => {
                                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                                const newResult = {};
                                Object.entries(record).forEach(([key, value]) => {
                                    if (!index) {
                                        Reflect.defineProperty(result, `${child.toString()}.${key}`, {
                                            value: value ?? chalk.bold('null'),
                                        });
                                    }
                                    else {
                                        Reflect.defineProperty(newResult, `${child.toString()}.${key}`, {
                                            value: value ?? chalk.bold('null'),
                                        });
                                    }
                                });
                                if (index) {
                                    subResults.push(newResult);
                                }
                            });
                        }
                    });
                }
                newResults.push(result, ...subResults);
                return newResults;
            }, []);
            return qr;
        }
        return queryResults;
    }
}
const SEPARATOR = ',';
const DOUBLE_QUOTE = '"';
const SHOULD_QUOTE_REGEXP = new RegExp(`[${SEPARATOR}${DOUBLE_QUOTE}${EOL}]`);
export class CsvReporter extends QueryReporter {
    constructor(data, columns) {
        super(data, columns);
    }
    display() {
        const attributeNames = this.massageRows();
        // begin output
        ux.log(attributeNames.map((name) => escape(name)).join(SEPARATOR));
        // explained why we need this below - foreach does not allow types
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        this.data.result.records.forEach((row) => {
            const values = attributeNames.map((name) => {
                // try get(row, name) first, then if it fails, default to row[name]. The default will happen in bulk cases.
                // the standard case returns {field:{nested: 'value'}}, while the bulk will return {field.nested: 'value'}
                const value = get(row, name, row[name]);
                if (isString(value)) {
                    return escape(value);
                    // if value is null, then typeof value === 'object' so check before typeof to avoid illegal csv
                }
                else if (value === null) {
                    return;
                }
                else if (typeof value === 'object') {
                    return escape(JSON.stringify(value));
                }
                return value;
            });
            ux.log(values.join(SEPARATOR));
        });
    }
    massageRows() {
        const fields = this.columns;
        const hasSubqueries = fields.some((field) => field.fieldType === FieldType.subqueryField);
        const hasFunctions = fields.some((field) => field.fieldType === FieldType.functionField);
        const attributeNames = [];
        if (fields) {
            this.logger.info(`Found fields ${JSON.stringify(fields.map((field) => `${typeof field}.${field.name}`))}`);
        }
        else {
            this.logger.info(`No fields found for query "${this.data.query}"`);
        }
        if (hasSubqueries || hasFunctions) {
            // If there are subqueries, we need to get the max child length for each subquery.
            const typeLengths = new Map();
            // For function fields, like avg(total).
            const aggregates = [];
            fields.forEach((field) => {
                if (field.fieldType === FieldType.subqueryField) {
                    typeLengths.set(field.name, 0);
                }
                if (field.fieldType === FieldType.functionField) {
                    aggregates.push(field);
                }
            });
            // Get max lengths by iterating over the records once
            this.data.result.records.forEach((result) => {
                [...typeLengths.keys()].forEach((key) => {
                    const record = get(result, key);
                    const totalSize = getNumber(record, 'totalSize');
                    if (!!totalSize && totalSize > (typeLengths.get(key) ?? 0)) {
                        typeLengths.set(key, totalSize);
                    }
                });
                // Aggregates are soql functions that aggregate data, like "SELECT avg(total)" and
                // are returned in the data as exprX. Aggregates can have aliases, like "avg(total) totalAverage"
                // and are returned in the data as the alias.
                if (aggregates.length > 0) {
                    for (let i = 0; i < aggregates.length; i++) {
                        const aggregate = aggregates[i];
                        if (!aggregate.alias) {
                            Reflect.set(result, aggregate.name, Reflect.get(result, `expr${i}`));
                        }
                    }
                }
            });
            fields.forEach((field) => {
                if (typeLengths.get(field.name)) {
                    for (let i = 0; i < (typeLengths.get(field.name) ?? 0); i++) {
                        attributeNames.push(`${field.name}.totalSize`);
                        (field.fields ?? []).forEach((subfield) => {
                            attributeNames.push(`${field.name}.records.${i}.${subfield.name}`);
                        });
                    }
                }
                else if (field.fieldType === FieldType.functionField) {
                    if (field.alias) {
                        attributeNames.push(field.alias);
                    }
                    else {
                        attributeNames.push(field.name);
                    }
                }
                else {
                    attributeNames.push(field.name);
                }
            });
        }
        else {
            attributeNames.push(...fields.map((field) => field.name));
        }
        return attributeNames;
    }
}
export class JsonReporter extends QueryReporter {
    constructor(data, columns) {
        super(data, columns);
    }
    // eslint-disable-next-line class-methods-use-this
    log() {
        return;
    }
    display() {
        ux.styledJSON({ status: 0, result: this.data.result });
    }
}
const prepColumns = (columns) => {
    const formattedColumns = {};
    columns
        .map((field) => field)
        .filter(isString)
        .map((field) => (formattedColumns[field] = {
        header: field.toUpperCase(),
        get: (row) => {
            // first test if key exists, if so, return value
            if (Reflect.has(row, field)) {
                return Reflect.get(row, field) ?? '';
            }
            else {
                // if not, try to find it query
                return get(row, field) ?? '';
            }
        },
    }));
    return formattedColumns;
};
/**
 * Escape a value to be placed in a CSV row. We follow rfc 4180
 * https://tools.ietf.org/html/rfc4180#section-2 and will not surround the
 * value in quotes if it doesn't contain the separator, double quote, or EOL.
 *
 * @param value The escaped value
 */
export const escape = (value) => {
    if (value && SHOULD_QUOTE_REGEXP.test(value)) {
        return `"${value.replace(/"/gi, '""')}"`;
    }
    return value;
};
export const getResultMessage = (jobInfo) => reporterMessages.getMessage('bulkV2Result', [
    jobInfo.id,
    capitalCase(jobInfo.state),
    jobInfo.numberRecordsProcessed,
    jobInfo.numberRecordsFailed,
]);
