// @salesforce/apex-node/lib/src/reporters/tapReporter.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TapReporter = void 0;
const elapsedTime_1 = require("../utils/elapsedTime");
class TapReporter {
    format(testResult, epilog) {
        const results = this.buildTapResults(testResult);
        const testPointCount = results.length;
        let out = '';
        out = out.concat(`1..${testPointCount}\n`);
        results.forEach((testPoint) => {
            out = out.concat(`${testPoint.outcome} ${testPoint.testNumber} ${testPoint.description}\n`);
            testPoint.diagnostics.forEach((s) => {
                out = out.concat(`# ${s}\n`);
            });
        });
        epilog?.forEach((c) => {
            out = out.concat(`# ${c}\n`);
        });
        return out;
    }
    buildTapResults(testResult) {
        const tapResults = [];
        testResult.tests.forEach((test, index) => {
            const testNumber = index + 1;
            const outcome = test.outcome === "Pass" /* ApexTestResultOutcome.Pass */ ? 'ok' : 'not ok';
            tapResults.push({
                testNumber,
                description: test.fullName,
                diagnostics: this.buildTapDiagnostics(test),
                outcome
            });
        });
        return tapResults;
    }
    buildTapDiagnostics(testResult) {
        const message = [];
        if (testResult.outcome !== 'Pass') {
            if (testResult.message) {
                const startsWithNewlineRegex = new RegExp(/^[/\r\n|\r|\n][\w]*/gim);
                if (startsWithNewlineRegex.test(testResult.message)) {
                    testResult.message.split(/\r\n|\r|\n/g).forEach((msg) => {
                        if (msg && msg.length > 0) {
                            message.push(msg.trim());
                        }
                    });
                }
                else {
                    message.push(testResult.message);
                }
            }
            else {
                message.push('Unknown error');
            }
            if (testResult.stackTrace) {
                testResult.stackTrace.split('\n').forEach((line) => {
                    message.push(line);
                });
            }
        }
        return message;
    }
}
exports.TapReporter = TapReporter;
__decorate([
    (0, elapsedTime_1.elapsedTime)()
], TapReporter.prototype, "format", null);
__decorate([
    (0, elapsedTime_1.elapsedTime)()
], TapReporter.prototype, "buildTapResults", null);
__decorate([
    (0, elapsedTime_1.elapsedTime)()
], TapReporter.prototype, "buildTapDiagnostics", null);
